✨ Scrum 53 be fast api 모델 생성

Ryan-OH opened the pull request:
## Explain this Pull Request 🙏

- Jira URL : https://9git-9git.atlassian.net/browse/SCRUM-53

## What has fixed? 🛠

변경 전 내용
(없음)
변경 후 내용
FastAPI 모델 추가, pr 템플릿, 기여 템플릿 추가
## Whether needed review 🖍

<!-- 마크다운의 빈 체크박스 문법은 [ ] / 채워진 체크박스 문법은 [x] 입니다. -->

- [x] 🙋🏻 리뷰가 필요합니다!
- [ ] 🙅🏻 리뷰가 필요하지 않습니다!

(이유 :FastAPI모델 코드 확인해 주세요! )

## Document🗒️ (Optional)

## Screenshot 📸 (Optional)

## Test Checklist ✅ (Optional)

<!-- 마크다운의 빈 체크박스 문법은 [ ] / 채워진 체크박스 문법은 [x] 입니다. -->

Ryan-OH committed:
✨ FastAPI모델, PR,기여 템플릿 추가

FastAPI모델, PR,기여 템플릿 추가

GideokKim commented:
@Ryan-OH 현재 `main` branch로 merge를 하게 설정되어 있네요 `dev`로 merge하게 바꿔주시면 좋을 것 같습니다

trivialcoding6 commented on app/models/base.py:
```diff
@@ -1,7 +1,23 @@
+    )
+
+    created_at: Mapped[datetime] = mapped_column(
+        DateTime(timezone=True), default=lambda: datetime.now(kst)
```
```suggestion
        DateTime(timezone=True), server_default=func.now()
```

default=lambda: datetime.now() 는 아래 링크와 사진 처럼 실제 디비에 추가되는 시간이 아닌 앱이 실행될때 호출되므로 실제 만들어진 시간과 차이가 있을 수 있어서 제가 제안한거로 수정해주시면 좋을 거 같습니다!
https://www.reddit.com/r/flask/comments/ypqk40/default_datetimenow_not_using_current_time/?rdt=43605

![스크린샷 2025-04-04 오전 11 18 24](https://github.com/user-attachments/assets/2be7c57e-4094-4fb0-b2f6-bb63e5fd4852)

Ryan-OH replied:
수정하겠습니다!

Ryan-OH replied:
수정했슴닷

trivialcoding6 commented on app/models/base.py:
```diff
@@ -1,7 +1,23 @@
+        DateTime(timezone=True), default=lambda: datetime.now(kst)
+    )
+    updated_at: Mapped[datetime] = mapped_column(
+        DateTime(timezone=True), default=lambda: datetime.now(kst)
```
```suggestion
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
```

update_at 같은 경우도 server_default로 수정하고  onupdate=func.now() 를 넣어주시면 수정될떄 자동으로 업데이트 된 시간으로 적용되니 참고 해주시면 좋을 거 같습니다!

Ryan-OH replied:
넵

trivialcoding6 commented on app/models/category.py:
```diff
@@ -0,0 +1,81 @@
+    today_notes: Mapped[list["TodayNotes"]] = relationship(
+        "TodayNotes", back_populates="category_progress", cascade="all, delete-orphan"
+    )
+    comprehensive_evaluations: Mapped[list["ComprehensiveEvaluations"]] = relationship(
+        "ComprehensiveEvaluations", back_populates="category_progress"
+    )
```
@Ryan-OH 
여기 gemini 가 남긴 코멘트는 이슈 없을까요?

Ryan-OH replied:
관계생각해봤을때 수정이 필요한것 같아서 우선 제미나이 추천대로 수정했습니다

trivialcoding6 commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,98 @@
+class MonthlyAchievements(Base):
+    __tablename__ = "monthly_achievements"
+
+    monthly_id: Mapped[str] = mapped_column(
```
혹시 여기는 자동 생성되는 id와 별개로 필요한 id 값일까요?

Ryan-OH replied:
그 제가 살펴보다가 그런류의 불필요한 id명명이 더 있어서 다 지워버렸습니다

trivialcoding6 commented:
@Ryan-OH 
gemini 가 남긴 코멘트들도 한번 확인해보시고 이슈 없으면 답글로 남겨주시면 좋을 거 같습니다!

Ryan-OH committed:
1. char(36) -> UUID(as_uuid=True)로 수정
2. table.table_id -> table.id 로 수정
3. 불필요한 코드 제거

GideokKim commented on app/models/base.py:
```diff
@@ -1,7 +1,23 @@
+    )
+
+    created_at: Mapped[datetime] = mapped_column(
+        DateTime(timezone=True), default=lambda: datetime.now(kst)
```
nullable 관련해서는 필요 없는지 잘 몰라서 여쭤봅니다!

@chaeengg @trivialcoding6 @Ryan-OH @trivialcoding6 

```suggestion
        DateTime(timezone=True), default=lambda: datetime.now(kst), nullable=False
```

GideokKim commented on app/models/base.py:
```diff
@@ -1,7 +1,23 @@
+        DateTime(timezone=True), default=lambda: datetime.now(kst)
+    )
+    updated_at: Mapped[datetime] = mapped_column(
+        DateTime(timezone=True), default=lambda: datetime.now(kst)
```
nullable 관련해서는 필요 없는지 잘 몰라서 여쭤봅니다!

@chaeengg @trivialcoding6 @Ryan-OH @trivialcoding6 

```suggestion
        DateTime(timezone=True), default=lambda: datetime.now(kst), nullable=False
```

Ryan-OH replied:
헌찬님이 주신 내용으로 수정했고 다시 커밋 올라갈 예정입니닷!
그 레코드가 업데이트 될때 자동으로 저장되는 방식이라 nullable옵션은 따로 추가 안해도 될것같아요

GideokKim reviewed (commented):
우선 `base.py` 관련해서 리뷰 남겼습니당

Ryan-OH commented on app/models/chat.py:
```diff
@@ -0,0 +1,84 @@
+        DateTime(timezone=True),
+        default=lambda: datetime.now(timezone.utc),
+        onupdate=lambda: datetime.now(timezone.utc),
+    )
```
날짜 관련해선 func.now 이용해서 수정완료

Ryan-OH commented on app/models/chat.py:
```diff
@@ -0,0 +1,84 @@
+        DateTime(timezone=True),
+        default=lambda: datetime.now(timezone.utc),
+        onupdate=lambda: datetime.now(timezone.utc),
+    )
```
날짜 관련해선 func.now 이용해서 수정완료

Ryan-OH commented on app/models/base.py:
```diff
@@ -0,0 +1,23 @@
+    )
+    updated_at: Mapped[datetime] = mapped_column(
+        DateTime(timezone=True), default=lambda: datetime.now(kst)
+    )
```
base 클래스에 대한 설명 추가 완료

Ryan-OH commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,98 @@
+    category_progress: Mapped["CategoryProgress"] = relationship(
+        "CategoryProgress", back_populates="comprehensive_evaluations"
+    )
+    # 관계 설정: ComprehensiveEvaluations → MonthlyAchievements (N:1 관계)
+    monthly_achievement: Mapped["MonthlyAchievements"] = relationship(
+        "MonthlyAchievements", back_populates="comprehensive_evaluations"
+    )
```
수정 완료

Ryan-OH commented on app/models/user.py:
```diff
@@ -0,0 +1,89 @@
+from enum import Enum as PyEnum
+from .base import Base
+from sqlalchemy.orm import relationship
+from category import Goals, TodayNotes, CategoryProgress
```
이건 수정 필요 없어보임

chaeengg commented on app/models/user.py:
```diff
@@ -0,0 +1,89 @@
+
+    # Character → User (N:1 관계)
+    user: Mapped["User"] = relationship(
+        "User", back_populates="characters", cascade="all, delete"
```
1. 여러 User가 동일한 Character를 가질 수 있으니 "Character → User (1:N) 관계" 로 수정하셔야 합니다
2. User가 삭제되더라도 Character는 그대로 유지되어야 하므로, 해당 테이블에서는 "cascade" 조건을 삭제하셔야 합니다

Ryan-OH commented on app/models/user.py:
```diff
@@ -0,0 +1,89 @@
+    job: Mapped[str] = mapped_column(String(100), nullable=False)
+
+    # User → UserCharacter (1:N 관계)
+    user_characters: Mapped[list["UserCharacter"]] = relationship(
+        "UserCharacter", back_populates="user"
+    )
+    # User → Character (1:N 관계)
+    characters: Mapped[list["Character"]] = relationship(
+        "Character", back_populates="user"
+    )
+    # User → Goals (1:N 관계)
+    goals: Mapped[list["Goals"]] = relationship("Goals", back_populates="user")
+
+    # User → TodayNotes (1:N 관계)
+    today_notes: Mapped[list["TodayNotes"]] = relationship(
+        "TodayNotes", back_populates="user"
+    )
+
+    # User → CategoryProgress (1:N 관계)
+    category_progress: Mapped[list["CategoryProgress"]] = relationship(
+        "CategoryProgress", back_populates="user"
+    )
```
여기에 대해서는 이걸 해야하는 이유가 있는지 잘 모르겠어서 공부해보겠습니다

GideokKim commented on app/models/user.py:
```diff
@@ -0,0 +1,86 @@
+    user_characters: Mapped[list["UserCharacter"]] = relationship(
+        "UserCharacter", back_populates="user"
+    )
+    # User → Character (1:N 관계)
+    characters: Mapped[list["Character"]] = relationship(
+        "Character", back_populates="user"
+    )
```
User와 Character 사이에 1대N 관계를 만들었는데, ERD에서는 그렇게 보이지 않네요. 혹시 설명 부탁드려도 될까요?

Ryan-OH replied:
그 제 생각이지만 1유저가 복수의 캐릭터를 소유하게 하기 위해서 그렇게 햇습니다...혹시 잘못된 관계면 없애버리겠습니다

Ryan-OH replied:
다시 조사해보니 관계가 반대로 가있었네요 수정하겠습니다!

chaeengg commented on app/models/category.py:
```diff
@@ -0,0 +1,81 @@
+    user_id: Mapped[str] = mapped_column(
+        String(36), ForeignKey("user.user_id"), nullable=False
+    )
+    category_name: Mapped[str] = mapped_column(
```
category_name에는 '파이썬/영어/운동' 값이 들어가기 때문에 primary key로 설정하면 안됩니다 ! 
base 파일에서 id 칼럼을 pk로 설정하셨으니 'primary_key=True'는 삭제하셔야 합니다

Ryan-OH committed:
수정완료

Ryan-OH committed:
수정의 수정
1. 칼럼명 복수형 -> 완료
2. Goals에 컬러 추가 -> 완료
3. 관계도 재정립 -> 완료
4. 임포트 정리 -> 완료
5. Enum통일 -> 완료
6. Demical항목 뒤에 default = 0.00 -> 완료

GideokKim commented on app/models/base.py:
```diff
@@ -1,7 +1,29 @@
+
+    created_at: Mapped[datetime] = mapped_column(
+        DateTime(timezone=True),
+        server_default=func.now(),
```
한줄로 있어도 되는 것 같습니다. `updated_at`이 더 길지만 분리가 안되어 있는걸 보면 이거도 한줄에 써도 될듯합니당.

```suggestion
        DateTime(timezone=True), server_default=func.now(),
```

Ryan-OH committed:
decimal -> default 0.00 수정완료

GideokKim commented on app/models/user.py:
```diff
@@ -0,0 +1,112 @@
+        "TodayNotes", back_populates="user"
+    )
+
+    # User → CategoryProgress (1:N 관계)
+    category_progress: Mapped[list["CategoryProgresses"]] = relationship(
+        "CategoryProgress", back_populates="user"
+    )
+
+    # User → Chats (1:N 관계)
+    chats: Mapped[list["Chats"]] = relationship("Chats", back_populates="user")
+
+    # User → Storages (1:N 관계)
+    storages: Mapped[list["Storages"]] = relationship("Storages", back_populates="user")
+    # User → Functions (1:N 관계)
+    functions: Mapped[list["Functions"]] = relationship(
+        "Functions", back_populates="user"
+    )
+
+    # User → RecommendedChallenge (1:N 관계)
+    recommended_challenges: Mapped[list["RecommendedChallenges"]] = relationship(
+        "RecommendedChallenge", back_populates="user"
+    )
```
순서를 이렇게 해주면 관련성 있는 것끼리 순서가 잡혀서 더 좋을 것 같습니당. 띄어쓰기도 신경써주세용.

```suggestion
    # User → Chats (1:N 관계)
    chats: Mapped[list["Chats"]] = relationship("Chats", back_populates="user")

    # User → Storages (1:N 관계)
    storages: Mapped[list["Storages"]] = relationship("Storages", back_populates="user")
    
    # User → Functions (1:N 관계)
    functions: Mapped[list["Functions"]] = relationship(
        "Functions", back_populates="user"
    )

    # User → CategoryProgress (1:N 관계)
    category_progress: Mapped[list["CategoryProgresses"]] = relationship(
        "CategoryProgress", back_populates="user"
    )
    
    # User → RecommendedChallenge (1:N 관계)
    recommended_challenges: Mapped[list["RecommendedChallenges"]] = relationship(
        "RecommendedChallenge", back_populates="user"
    )
```

GideokKim commented on app/models/user.py:
```diff
@@ -0,0 +1,112 @@
+    )
+    character_id: Mapped[UUID] = mapped_column(
+        UUID(as_uuid=True), ForeignKey("character.id"), nullable=False
+    )
+    # UserCharacter → User (N:1 관계)
+    user: Mapped["User"] = relationship(
+        "User", back_populates="user_characters", cascade="all, delete"
+    )
+    # UserCharacter → Character (N:1 관계)
+    character: Mapped["Character"] = relationship(
+        "Character", back_populates="user_characters", cascade="all, delete"
+    )
```
띄어쓰기 추가입니당

```suggestion
    )
    
    # UserCharacter → User (N:1 관계)
    user: Mapped["User"] = relationship(
        "User", back_populates="user_characters", cascade="all, delete"
    )
    
    # UserCharacter → Character (N:1 관계)
    character: Mapped["Character"] = relationship(
        "Character", back_populates="user_characters", cascade="all, delete"
    )
```

trivialcoding6 commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,100 @@
+    category_name: Mapped[CategoryEnum] = mapped_column(
+        Enum(CategoryEnum, name="category_enum"), nullable=False
+    )
+    progress_rate: Mapped[DECIMAL] = mapped_column(DECIMAL(5, 2), defalt=0.00)
```
@Ryan-OH 
별건 아니지만 여기 default 오타가 있는거 같습니닷

Ryan-OH replied:
앗 수정하겠습니다

GideokKim commented on app/models/user.py:
```diff
@@ -0,0 +1,112 @@
+
+
+# Character 테이블
+class Character(Base):
```
character name이 없는 것 같습니당

Ryan-OH replied:
앗 이부분을 놓치다니 가독성측면도 고려해서 수정하겠습니다

GideokKim reviewed (commented):
일단 `user.py`에 대한 코드 리뷰 먼저드립니당

trivialcoding6 commented on app/models/user.py:
```diff
@@ -0,0 +1,112 @@
+    __tablename__ = "character"
+
+    user_id: Mapped[UUID] = mapped_column(
+        UUID(as_uuid=True), ForeignKey("user.id"), nullable=False
```
그 다대다 관계에서
유저 - 유저 캐릭터 - 캐릭터 요런 관계이면 여기에 foreignKey("user_character.user_id") 로 설정해야하지 않나 싶네요

Ryan-OH replied:
아하...그부분은 전혀 생각못했습니닷

GideokKim commented on app/models/chat.py:
```diff
@@ -0,0 +1,88 @@
+from .base import Base
+from .user import User
+from enum import Enum as PyEnum
+from category import CategoryEnum
```
위에 `base`나 `user`처럼 import하는 형식이 똑같아야 할 것 같아요!

```suggestion
from .category import CategoryEnum
```

GideokKim commented on app/models/chat.py:
```diff
@@ -0,0 +1,88 @@
+        UUID(as_uuid=True), ForeignKey("functions.id"), nullable=False
+    )
+    storage_title: Mapped[str] = mapped_column(String(255), nullable=False)
+    storage_description: Mapped[str] = mapped_column(Text, nullable=True)
```
```suggestion
    storage_description: Mapped[str] = mapped_column(Text)
```

GideokKim replied:
@Ryan-OH `storage_description`는 null 값이 가능합니다

Ryan-OH replied:
넵 바로수정하겠습니다

GideokKim commented on app/models/chat.py:
```diff
@@ -0,0 +1,88 @@
+    __tablename__ = "functions"
+
+    function_name: Mapped[str] = mapped_column(String(255), nullable=False)
+    function_description: Mapped[str] = mapped_column(Text, nullable=True)
```
```suggestion
    function_description: Mapped[str] = mapped_column(Text)
```

GideokKim replied:
@Ryan-OH 수정되지 않아서 unresolve합니다

GideokKim commented on app/models/category.py:
```diff
@@ -0,0 +1,104 @@
+        "User", back_populates="category_progress", cascade="all, delete"
+    )
+
+    # CategoryProgresses -> Goals (N : 1 관계)
+    goals: Mapped[list["Goals"]] = relationship(
+        "Goals", back_populates="category_progress", cascade="all, delete"
+    )
```
주석이랑 코드가 일치하지 않는 것 같습니당 ERD 보시고 수정하시거나 멘트 남겨주세용

Ryan-OH replied:
관계수정하고 기덕님이 주신대로 픽스했습니닷!

GideokKim replied:
@Ryan-OH 실제 수정되지 않았습니다 현재 코드에 따로 추가 멘트드렸으니 확인 부탁드립니당

GideokKim commented on app/models/category.py:
```diff
@@ -0,0 +1,104 @@
+    completed_goal: Mapped[int] = mapped_column(Integer, nullable=False)
+    progress_rate: Mapped[DECIMAL] = mapped_column(DECIMAL(5, 2), default=0.00)
+
+    user: Mapped["User"] = relationship(
+        "User", back_populates="category_progress", cascade="all, delete"
+    )
+
+    # CategoryProgresses -> Goals (N : 1 관계)
+    goals: Mapped[list["Goals"]] = relationship(
+        "Goals", back_populates="category_progress", cascade="all, delete"
+    )
+
+    # CategoryProgesses -> ComprehensiveEvaluations (1 : N 관계)
+    comprehensive_evaluations: Mapped[list["ComprehensiveEvaluations"]] = relationship(
+        "ComprehensiveEvaluations", back_populates="category_progress"
+    )
+
+    # CategoryProgesses -> RecommandedChallenges (1 : N 관계)
+    recommended_challenges: Mapped[list["RecommendedChallenges"]] = relationship(
+        "RecommendedChallenges", back_populates="category_progress"
+    )
```
관계 순서 정리할 때 1대N부터 적어주시고 N대1 그 다음에 적어주세용

Ryan-OH replied:
이건 아까 토의한대로 USER먼저 정리하겠습니닷

GideokKim commented on app/models/category.py:
```diff
@@ -0,0 +1,104 @@
+    is_completed: Mapped[bool] = mapped_column(Boolean, default=False)
+    is_repeat: Mapped[bool] = mapped_column(Boolean, nullable=False)
+
+    user: Mapped["User"] = relationship(
+        "User", back_populates="goals", cascade="all, delete"
+    )
+
+    # Goals -> CategoryProgresses (1 : N 관계)
+    category_progress: Mapped["CategoryProgresses"] = relationship(
+        "CategoryProgresses", back_populates="goals"
+    )
```
`Goals -> CategoryProgresses (1 : N 관계)`이면 코드 이렇게 수정해야 하지 않나 싶은데 맞나용?
그리고 위에 언급했듯이 1대N부터 적고 N대1 적는걸로 순서 변경했습니다

```suggestion
    # Goals -> CategoryProgresses (1 : N 관계)
    category_progress: Mapped[list[""CategoryProgresses"]] = relationship(
        "CategoryProgresses", back_populates="goals"
        
    # Goals -> User (N : 1 관계)
    user: Mapped["User"] = relationship(
        "User", back_populates="goals", cascade="all, delete"
    )
    )
```

Ryan-OH replied:
넵 수정완료했고 순서는 토의한대로 진행하겠습니닷

GideokKim commented on app/models/category.py:
```diff
@@ -0,0 +1,104 @@
+
+
+# 색상 Enum 클래스 정의
+class CategoryColor(PyEnum):
```
```suggestion
class CategoryColorEnum(PyEnum):
```

Ryan-OH replied:
네 수정했습ㄴ다

GideokKim commented on app/models/category.py:
```diff
@@ -0,0 +1,104 @@
+from evaluation import RecommendedChallenges
+
+
+# 카테고리 Enum 클래스 정의
+class CategoryEnum(PyEnum):
```
ERD 보면 `category_name`인데 이거도 아래처럼 수정해야 하지 않나싶습니당

```suggestion
# 카테고리 이름 Enum 클래스 정의
class CategoryNameEnum(PyEnum):
```

GideokKim replied:
@Ryan-OH 주석 수정이 누락되어서 다시 unresolve 처리합니다

Ryan-OH replied:
처리완료했습니다

GideokKim commented on app/models/category.py:
```diff
@@ -0,0 +1,104 @@
+    )
+    title: Mapped[str] = mapped_column(String(255), nullable=False)
+    content: Mapped[str] = mapped_column(Text, nullable=False)
+
```
```suggestion

# TodayNotes -> User (1 : N 관계)
```

GideokKim reviewed (commented):
`chat.py`랑 `category.py`에 대한 코드 리뷰입니당

GideokKim commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,100 @@
+from .base import Base
+from .user import User
+from enum import Enum as PyEnum
+from category import CategoryProgresses, CategoryEnum
```
`CategoryProgresses`가 필요한가요? relationship을 위해 필요한거라면 그동안 작성된 모든 모델 코드에 다 relationship을 위한 import가 들어가야 하고 만약 필요 없다면 여기만 지우시면 됩니당

```suggestion
from .category import CategoryNameEnum
```

Ryan-OH replied:
이건 RecommendedChalenge 테이블에 관계설정할때  필요해서 했어요

GideokKim commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,100 @@
+from uuid import UUID
+from .base import Base
+from .user import User
+from enum import Enum as PyEnum
```
안쓰는 것 같은데 확인 부탁드립니당

```suggestion
```

Ryan-OH replied:
이건 지웠습니다

GideokKim commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,100 @@
+
+
+# RecommendedChallenge 테이블 (추천 챌린지)
+class RecommendedChallenges(Base):
```
ERD에 있는 `progress_id`에 대해 언급이 없는데 Base를 상속 받아서 별칭 처리 같은 추가적인 처리는 필요 없나요?

Ryan-OH replied:
ERD수정 후에 변동된내용이라 적용시키고 해결 누르겠습니다

Ryan-OH replied:
수정했슴닷

GideokKim commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,100 @@
+    challenge_difficulty: Mapped[str] = mapped_column(String(20), nullable=True)
+    challenge_suggestion: Mapped[str] = mapped_column(Text, nullable=True)
+
+    # 관계 설정: RecommendedChallenge → User (N:1 관계) / User가 삭제되면 전부 삭제
+    user: Mapped["User"] = relationship(
+        "User", back_populates="recommended_challenges", cascade="all, delete"
+    )
+
+    # 관계 설정: RecommendedChallenge → CategoryProgress (N:1 관계)
+    category_progress: Mapped["CategoryProgresses"] = relationship(
+        "CategoryProgresses",
+        back_populates="recommended_challenges",
+        cascade="all, delete",
+    )
```
User table과 관계가 있는건가요? ERD 상에서는 User table이 아닌 CategoryProgresses table이랑만 관련이 있어보여서요!

```suggestion
    # 관계 설정: RecommendedChallenge → User, CategoryProgress (N:1 관계)
    user: Mapped["User"] = relationship(
        "User", back_populates="recommended_challenges", cascade="all, delete"
    )
    category_progress: Mapped["CategoryProgresses"] = relationship(
        "CategoryProgresses",
        back_populates="recommended_challenges",
        cascade="all, delete",
    )
```

Ryan-OH replied:
user가 사라지면 recommended 도 할 필요가 없어지니 그래서 결국 들어가는게 맞다고 들어서 관계는 있는것 같습니닷!

GideokKim replied:
`User` table과 관계를 짓지 않아도 
`User` table이 삭제되면 `CategoryProgress` table이 삭제되고
`CategoryProgress` table이 삭제되면 `RecommendedChallenge` table이 삭제되는 연쇄 구조로 되어 있는 것 같은데 
`User` table과 관계가 있어야 할까요?

이 부분은 코드랑 DB에 익숙하신 @trivialcoding6 @chaeengg 의 의견을 들어보고 싶습니당

trivialcoding6 replied:
@GideokKim 
네 기덕님 말씀대로 그렇게 연쇄 참조되어 있으면 RecommendedChallenge 에서는 user를 직접 참조하지 않아도 될거같습니다!

Ryan-OH replied:
그럼 우선 이 코멘트대로 User테이블과 RecommendedChallenge 테이블의 직접적인 관계는 끊어놓겠습니다

chaeengg commented on app/models/category.py:
```diff
@@ -1,33 +1,60 @@
-    __tablename__ = "category_progress"
+class CategoryProgresses(Base):
+    __tablename__ = "category_progresses"
 
```
ERD에 Goal 테이블에서 참조한 goal_id 추가했습니다! 코드에도 해당 칼럼 추가 부탁드립니당

Ryan-OH replied:
칼럼 추가했슴닷

GideokKim commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,100 @@
+
+    user_id: Mapped[UUID] = mapped_column(
+        UUID(as_uuid=True), ForeignKey("user.id"), nullable=False
+    )  # User를 참조
```
User table에 있는 user.id를 참조하는건지 CategoryProgresses table에 있는 user.id를 참고하는지 제가 잘 몰라서 여쭤봅니당.

ERD에서는 CategoryProgresses table이랑만 연결되어 있는 것 같아서요!

Ryan-OH replied:
아 유저아이디는 CategoryProgresses 여기서 따오는게 맞아서 재정의 했습니다

GideokKim commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,100 @@
+    overall_achievement_rate: Mapped[DECIMAL] = mapped_column(
+        DECIMAL(5, 2), default=0.00
+    )
+    evaluation_text: Mapped[str] = mapped_column(Text, nullable=True)
```
```suggestion
    evaluation_text: Mapped[str] = mapped_column(Text)
```

GideokKim replied:
잘못 수정되어 unresolve 처리합니다

ERD 보시면 null이 가능한 값인데 null이 불가능하도록 수정하셨어요

Ryan-OH replied:
이것도 수정 되어있는데 재커밋할때 다시 살펴볼게요

GideokKim commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,100 @@
+    strength_achievement_rate: Mapped[DECIMAL] = mapped_column(
+        DECIMAL(5, 2), default=0.00
+    )
+    strength_text: Mapped[str] = mapped_column(Text, nullable=True)
```
```suggestion
    strength_text: Mapped[str] = mapped_column(Text)
```

GideokKim replied:
동일하게 잘못 수정되어 다시 unresolve로 변경합니다.

chaeengg commented on app/models/evaluation.py:
```diff
@@ -1,31 +1,23 @@
-    __tablename__ = "recommended_challenge"
+class RecommendedChallenges(Base):
+    __tablename__ = "recommended_challenges"
 
```
ERD에 Category_Progresses 테이블에서 참조한 progress_id 칼럼 추가했습니다 해당 코드에도 추가 부탁드립니당

Ryan-OH replied:
추가했습니다

GideokKim commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,100 @@
+    )
+    strength_text: Mapped[str] = mapped_column(Text, nullable=True)
+    improvement_category: Mapped[str] = mapped_column(String(100), nullable=False)
+    improvement_achievement_rate: Mapped[DECIMAL] = mapped_column(DECIMAL(5, 2))
```
```suggestion
    improvement_achievement_rate: Mapped[DECIMAL] = mapped_column(DECIMAL(5, 2), default=0.00)
```

GideokKim replied:
default 값이 없어서 다시 unresolve 처리합니다.

Ryan-OH replied:
오타수정했습니다

Ryan-OH replied:
아 정정완료했습니다

GideokKim commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,100 @@
+    # 관계 설정: MonthlyAchievements → User
+    user: Mapped["User"] = relationship("User", back_populates="monthly_achievements")
+
+    # 관계 설정 : MonthlyAchievements -> ConprehensiveEvaluations (1 : N 관계)
```
```suggestion
    # 관계 설정 : MonthlyAchievements -> ConprehensiveEvaluations (1:N 관계)
```

GideokKim commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,100 @@
+    completed_goal: Mapped[int] = mapped_column(Integer, nullable=False)
+    progress_rate: Mapped[DECIMAL] = mapped_column(DECIMAL(5, 2), default=0.00)
+
+    # 관계 설정: MonthlyAchievements → User
```
```suggestion
    # 관계 설정: MonthlyAchievements → User (N:1 관계)
```

GideokKim reviewed (commented):
`evaluation.py` 코드 리뷰입니다!

Ryan-OH committed:
찐찐찐찐 최종본입니다

Ryan-OH committed:
이와중에 임포트에 function을 안지워서 지웠습니다

GideokKim commented on app/models/category.py:
```diff
@@ -0,0 +1,103 @@
+    users: Mapped["User"] = relationship(
+        "User", back_populates="category_progresses", cascade="all, delete"
+    )
+    goals: Mapped[list["Goal"]] = relationship(
+        "Goal", back_populates="category_progresses", cascade="all, delete"
+    )
```
관계가 이렇게 되는것 같은데 확인 부탁드립니당 

```suggestion
    goals: Mapped["Goal"] = relationship(
        "Goal", back_populates="category_progresses", cascade="all, delete"
    )
```

Ryan-OH replied:
이부분은 제가 약간 헷갈리는데 리스트를 달아주면 하나의 카테고리가 여러개의 목표를 가질수 있게되는데,
하나의 카테고리가 하나의 목표를 바라보는게 적절한거 같긴해서 일단 수정은 해놓고 질문 남겨봅니다
@trivialcoding6 @chaeengg

trivialcoding6 replied:
@Ryan-OH 
목표와 진행률 테이블을 말씀 하신거로 보이는데요! 기덕님이 제안하신게 리스트가 어니라 1:1 관계가 되도록 제안주신거 같은데 그럼 동준님이 말해주신거랑 일치하는거 아닌가요?!

Ryan-OH replied:
결국 차이는 categoryprogress가 여러개의 goal을 가질수 있냐 없나의 관점에서 list를 다느냐 안다느냐의 차이인것 같은데 확인부탁드립니다 우선 수정은 list를 제외하였습니다

trivialcoding6 replied:
@Ryan-OH 
제가 알기로는 goals -> 영어 코딩 운동 이거로 알고있는데 각 목표별로 진행률이 여러개일수있나요..!? 혹시 제가 잘못 파악하고 있는거면 정정 부탁드립니다🙏🙏

cc @GideokKim @chaeengg

GideokKim replied:
저는 CategoryProgress → Goal가 N : 1 관계를 가진다고 ERD에서 봤는데 list가 없는게 N : 1이고 list가 있으면 1 : N 관계로 묶인다고 코드상 이해했어요.

만약 list가 없는 표현이 1:1 관계를 가진다고 하면 지금 코드 전체에 걸쳐 user table과의 관계가 N : 1이 아닌 1:1이 된다는 의미이므로 ERD랑 다르게 코드를 작성한게 되는데 어떤게 맞는건가요?

정리하자면 아래 코드가 N대1을 의미하는지 1대1을 의미하는지 알려주세요!

```python
 users: Mapped["User"] = relationship(
        "User", back_populates="category_progresses", cascade="all, delete"
    )
```

Ryan-OH replied:
기덕님이 이해하신계 맞아요 아래코드는 N:1관계고   list가 없는게 N : 1이고 list가 있으면 1 : N 관계로 묶인다가 맞는것 같습니다

GideokKim replied:
@trivialcoding6 ERD 상에는 CategoryProgress → Goal가 N : 1의 관계를 가진다고 ERD에 작성되어 있는거로 이해해서 코드 리뷰를 그렇게 했고 실제로 그게 가능한 건지는 생각하지 않았습니당 저는 코드와 ERD가 일치하는지만 체크하면서 코드 리뷰를 했어요! 말씀하신 부분 생각하니 저도 헷깔리는데 이 부분은 따로 정리가 필요할 것 같고 지금은 ERD랑 일치시키는 것만 목표로 코드 리뷰를 하겠습니당

@Ryan-OH 넵 그렇다면 지금 코드가 ERD와 일치한다고 보여서 이렇게 작성하는게 맞다고 생각됩니당

trivialcoding6 replied:
@GideokKim 
아 넵넵!! 저도 생각해보니 채은님한테 동일하게 여쭤봐서 답변을 받았던거 같은데 기억이 잘 안나네요 🥲 ERD 상으론 맞는거 같아서 저도 이의 없습니다!

cc @Ryan-OH @chaeengg

Ryan-OH replied:
확인했고 기존에 수정했던대로 (기덕님이 제시하신 코드로) 냅두겠습니닷!

Ryan-OH committed:
chat.py nullable값 수정

GideokKim commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,97 @@
+class RecommendedChallenge(Base):
+    __tablename__ = "recommended_challenges"
+    progress_id = Mapped[UUID] = mapped_column(
+        UUID(as_uuid=True), ForeignKey("category_progresses.id"), nullable=False
+    )
+    user_id: Mapped[UUID] = mapped_column(
+        UUID(as_uuid=True), ForeignKey("CategoryProgresses.user_id"), nullable=False
```
아래처럼 `category_progresses`로 통일해야하는 것 같은데 `CategoryProgresses`로 써도 되는건가요?

```suggestion
        UUID(as_uuid=True), ForeignKey("category_progresses.id"), nullable=False
    )
    user_id: Mapped[UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("category_progresses.id.user_id"), nullable=False
```

Ryan-OH replied:
저건 이제 해당 클래스의 고유 id 가져올때는 테이블명.id로 가져오고,  다른 키 참조할때는 클래스명.해당키명 을 가져오는거라고 알고있어서 확인 부탁드립니다
@trivialcoding6 @chaeengg

trivialcoding6 replied:
![Screenshot_20250405_122351_Samsung Internet](https://github.com/user-attachments/assets/d2031201-b5a1-4882-8796-425bc3ba69a2)
 코드 보면 다른곳도 테이블명으로 되어있는거 같고 참조키 설정에는 테이블 명이 들어가는게 맞는거 같은데 한번 다시 확인 부탁드립니다!

Ryan-OH replied:
제가 잘못 알고있ㄷ건것 같습니다 수정할게욥

GideokKim commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,97 @@
+        Enum(CategoryNameEnum, name="category_name_enums"), nullable=False
+    )
+    progress_rate: Mapped[DECIMAL] = mapped_column(DECIMAL(5, 2), default=0.00)
+    challenge_task: Mapped[str] = mapped_column(Text, nullable=True)
+    challenge_duration: Mapped[str] = mapped_column(String(50), nullable=True)
+    challenge_difficulty: Mapped[str] = mapped_column(String(20), nullable=True)
+    challenge_suggestion: Mapped[str] = mapped_column(Text, nullable=True)
```
```suggestion
    challenge_task: Mapped[str] = mapped_column(Text)
    challenge_duration: Mapped[str] = mapped_column(String(50))
    challenge_difficulty: Mapped[str] = mapped_column(String(20))
    challenge_suggestion: Mapped[str] = mapped_column(Text)
```

Ryan-OH replied:
뭐지 제 코드에는 이거 다 지워져있는데 커밋할때 누락되었나봅니다 다시 커밋할때는 잘 바뀌었는지 확인해보겠습니다

GideokKim reviewed (commented):
수정이 안된 부분 전부 unresolve 처리하였고 추가 코멘트 달았습니당

Ryan-OH committed:
찐최종
누락된값 전부 재커밋 시키겠습니다
User와 RecommendedChallenges의 관계를 끊었습니다
goal과 CategoryProgress의 관계에서 list를 제거하였습니다

Ryan-OH committed:
category_progresses.user_id 변경

chaeengg commented on app/models/category.py:
```diff
@@ -8,14 +8,14 @@
+# CategoryColorEnum 클래스 정의
 class CategoryColorEnum(PyEnum):
     CODING = "#FDA63A"  # 호박색
     ENGLISH = "#6C88C4"  # 인디고 블루
```
Coding과 English의 color는 지정되어있는데 Workout의 색깔이 정의되지 않았습니다
#556B2F이 운동 color로 erd에 있는데 해당 부분 수정 부탁드립니다

Ryan-OH replied:
이 아래쪽줄에 바로 있지 않나욤?

chaeengg replied:
엇 죄송합니다,,,

Ryan-OH committed:
CategoryProgress -> Goals 관계 재 확인후 관계 재설정

Ryan-OH committed:
오타 수정

Ryan-OH committed:
nullable 재수정

GideokKim commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,96 @@
+    improvement_achievement_rate: Mapped[DECIMAL] = mapped_column(
+        DECIMAL(5, 2), default=0.00
+    )
+    improvement_text: Mapped[str] = mapped_column(Text, nullable=False)
```
ERD 상에서는 null이 가능한 값입니당

```suggestion
    improvement_text: Mapped[str] = mapped_column(Text)
```

Ryan-OH replied:
수정하였습니다

GideokKim commented on app/models/user.py:
```diff
@@ -0,0 +1,86 @@
+        Enum(GenderEnum, name="gender_enum"), nullable=False
+    )
+    age: Mapped[int] = mapped_column(Integer)
+    job: Mapped[str] = mapped_column(String(100), nullable=False)
```
ERD에서는 null이 가능한 값입니당

```suggestion
    job: Mapped[str] = mapped_column(String(100))
```

Ryan-OH replied:
수정하였고 ERD상에서 NN되있는거 일괄적으로 nullable=False처리하였습니다(통일성)

GideokKim commented on app/models/user.py:
```diff
@@ -0,0 +1,86 @@
+    job: Mapped[str] = mapped_column(String(100), nullable=False)
+
+    # User → UserCharacter, Goal, TodayNote, CategoryProgress, Chat, Storage, ComprehensiveEvaluation, MonthlyAchievement (1:N 관계)
+    user_characters: Mapped[list["UserCharacter"]] = relationship(
+        "UserCharacter", back_populates="users"
+    )
+    goals: Mapped[list["Goal"]] = relationship("Goal", back_populates="users")
+    today_notes: Mapped[list["TodayNote"]] = relationship(
+        "TodayNote", back_populates="users"
+    )
+    category_progresses: Mapped[list["CategoryProgress"]] = relationship(
+        "CategoryProgress", back_populates="users"
+    )
+    chats: Mapped[list["Chat"]] = relationship("Chat", back_populates="users")
+    storages: Mapped[list["Storage"]] = relationship("Storage", back_populates="users")
+    comprehensive_evaluations: Mapped[list["ComprehensiveEvaluation"]] = relationship(
+        "ComprehensiveEvaluation", back_populates="users"
+    )
+    monthly_achievements: Mapped[list["MonthlyAchievement"]] = relationship(
+        "MonthlyAchievement", back_populates="users"
+    )
```
여기 혹시 단수로 써야하는거 아닌가용? 저도 잘 몰라서 여쭤봅니당

```suggestion
    user_characters: Mapped[list["UserCharacter"]] = relationship(
        "UserCharacter", back_populates="user"
    )
    goals: Mapped[list["Goal"]] = relationship("Goal", back_populates="users")
    today_notes: Mapped[list["TodayNote"]] = relationship(
        "TodayNote", back_populates="user"
    )
    category_progresses: Mapped[list["CategoryProgress"]] = relationship(
        "CategoryProgress", back_populates="user"
    )
    chats: Mapped[list["Chat"]] = relationship("Chat", back_populates="user")
    storages: Mapped[list["Storage"]] = relationship("Storage", back_populates="user")
    comprehensive_evaluations: Mapped[list["ComprehensiveEvaluation"]] = relationship(
        "ComprehensiveEvaluation", back_populates="user"
    )
    monthly_achievements: Mapped[list["MonthlyAchievement"]] = relationship(
        "MonthlyAchievement", back_populates="user"
    )
```

Ryan-OH replied:
이건 User Class의 테이블 명이 users여서 users 맞는것 같습니다

GideokKim replied:
넵 그럼 이건 resolve 처리하고 다른 곳에 코멘트로 하나 더 여쭤보겠습니당

GideokKim reviewed (commented):
관계 설정 관련 질문이랑 nullable 관련 추가 리뷰입니당

Ryan-OH committed:
nullable관련 ERD상에 있는걸로 통일성 맞췄습니다

GideokKim commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,100 @@
+    improvement_text: Mapped[str] = mapped_column(Text)
+
+    # ComprehensiveEvaluation → User, MonthlyAchievement, CategoryProgress (N:1 관계)
+    user: Mapped["User"] = relationship(
+        "User", back_populates="comprehensive_evaluations", cascade="all, delete"
+    )
+    monthly_achievements: Mapped["MonthlyAchievement"] = relationship(
```
여기 `user:`는 단수고 `monthly_achievements:`는 복수인데 이건 상관 없는건가용?

```suggestion
    user: Mapped["User"] = relationship(
        "User", back_populates="comprehensive_evaluations", cascade="all, delete"
    )
    monthly_achievements: Mapped["MonthlyAchievement"] = relationship(
```

Ryan-OH replied:
자료 찾아보니 back_populates="user"에 해당하는 것도 속성명이고  user:부분에 해당하는것도 속성명이라  두개가 같다면 된다고하네요 그러면 그냥 user로 되어있는거 전부 users로 맞출까요? 앞 뒤 전부

GideokKim replied:
@Ryan-OH 일단 이건 resolve 처리하고 아래에 헌찬님이 관련 내용 또 언급하셨으니 거기서 헌찬님과 논의 후 처리하는거로 하시죠!

 @trivialcoding6

GideokKim reviewed (commented):
이 부분 궁금합니당

trivialcoding6 commented on app/models/category.py:
```diff
@@ -0,0 +1,101 @@
+    progress_rate: Mapped[DECIMAL] = mapped_column(DECIMAL(5, 2), default=0.00)
+
+    # CategoryProgress → User (N : 1 관계)
+    users: Mapped["User"] = relationship(
```
@Ryan-OH 
기덕님도 아래 남겨주신거 같은데 N : 1 관계 (user가 1) 이라면 여기 단수가 맞는거 같네요
```suggestion
    user: Mapped["User"] = relationship(
```

Ryan-OH replied:
현재 테이블명은 users로 되어있고, 앞부분은 속성명인데 그러면 user부분 속성명 전부 단수처리 할까요?

trivialcoding6 replied:
@Ryan-OH 
네네!! 번거롭겠지만 그렇게 하는게 맞는거 같습니다 ㅠㅠ users 테이블 명에 상관없이 속성이름이므로 속성이 단수라면 단수로 하는게 맞는거 같습니닷..!

trivialcoding6 commented on app/models/category.py:
```diff
@@ -0,0 +1,101 @@
+    )
+
+    # CategoryProgress → ComprehensiveEvaluation, RecommendedChallenge, Goal (1 : N 관계)
+    comprehensive_evaluations: Mapped[list["ComprehensiveEvaluation"]] = relationship(
```
```suggestion
    comprehensive_evaluations: Mapped[List["ComprehensiveEvaluation"]] = relationship(
```
@Ryan-OH 
조금 뒷북같지만 죄송합니다 ㅠㅠ sqlalchemy 공식문서 좀 보다가 발견했는데 관계에서 list인 부분들 전체적으로 typing 의 List로 안해도 문제 없을까요?

링크: https://docs.sqlalchemy.org/en/20/orm/quickstart.html#declare-models
<img width="857" alt="스크린샷 2025-04-06 오후 2 44 46" src="https://github.com/user-attachments/assets/1488798c-65ce-4dda-bb16-90ff15483fb9" />

Ryan-OH replied:
기존 코드 (list[...]) → 작동 가능하지만, IDE 자동완성 및 타입 검사에서 조금 덜 명확할 수 있음
 개선 코드 (List[...]) → IDE 및 타입 검사에서 더 명확하게 인식되므로 코드 가독성이 좋아짐
바꾸는 쪽이 더 좋은것 같긴합니다 바꿀까요?

GideokKim reviewed (approved):
LGTM

Ryan-OH committed:
list -> List로 변경
속성명 -> users -> user
나머지 속성명은 전부 복수처리(통일성)

Ryan-OH committed:
오타수정

Ryan-OH committed:
'(속성명):  '관계 보고 복수 단수 수정
back_populates = '클래스'명으로 전부 정정

trivialcoding6 commented:
LGTM

GideokKim commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,101 @@
+        back_populates="comprehensive_evaluations",
+        cascade="all, delete",
+    )
+    category_progresse: Mapped["CategoryProgress"] = relationship(
```
이미 approve 눌렀지만 눈에 보여서 코멘트 남깁니당

```suggestion
    category_progress: Mapped["CategoryProgress"] = relationship(
```

Ryan-OH replied:
앗..! 역시..코딩은 한글자 차이...죄송합니다 수정완료했습니다

Ryan-OH committed:
오타수정

codingorijazz commented on app/models/chat.py:
```diff
@@ -0,0 +1,88 @@
+
+    # Storage → User, Function (N:1 관계)
+    user: Mapped["User"] = relationship(
+        "User", back_populates="storages", cascade="all, delete"
```
```suggestion
        "User", back_populates="storages"
```

codingorijazz commented on app/models/chat.py:
```diff
@@ -0,0 +1,88 @@
+        "User", back_populates="storages", cascade="all, delete"
+    )
+    function: Mapped["Function"] = relationship(
+        "Function", back_populates="storages", cascade="all, delete"
```
```suggestion
        "Function", back_populates="storages"
```

codingorijazz commented on app/models/chat.py:
```diff
@@ -0,0 +1,88 @@
+    )
+
+    # Storage → Chat (1:N 관계)
+    chats: Mapped[List["Chat"]] = relationship("Chat", back_populates="storages")
```
```suggestion
    chats: Mapped[List["Chat"]] = relationship("Chat", back_populates="storage", cascade="all, delete")
```

codingorijazz replied:
앗 제가 스페이스 두칸을 눌러버렸네요.

codingorijazz commented on app/models/chat.py:
```diff
@@ -0,0 +1,88 @@
+
+    # Chat → User, Storage, Function (N:1 관계)
+    user: Mapped["User"] = relationship(
+        "User", back_populates="chats", cascade="all, delete"
```
```suggestion
        "User", back_populates="chats"
```

codingorijazz commented on app/models/chat.py:
```diff
@@ -0,0 +1,88 @@
+        "User", back_populates="chats", cascade="all, delete"
+    )
+    storage: Mapped["Storage"] = relationship(
+        "Storage", back_populates="chats", cascade="all, delete"
```
```suggestion
        "Storage", back_populates="chats"
```

codingorijazz commented on app/models/chat.py:
```diff
@@ -0,0 +1,88 @@
+        "Storage", back_populates="chats", cascade="all, delete"
+    )
+    function: Mapped["Function"] = relationship(
+        "Function", back_populates="chats", cascade="all, delete"
```
```suggestion
        "Function", back_populates="chats"
```

codingorijazz commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,101 @@
+    # RecommendedChallenge → CategoryProgress (N:1 관계)
+    category_progress: Mapped["CategoryProgress"] = relationship(
+        "CategoryProgress",
+        back_populates="recommended_challenges",
+        cascade="all, delete",
```
```suggestion
        back_populates="recommended_challenges"
```

codingorijazz commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,101 @@
+    )
+    improvement_text: Mapped[str] = mapped_column(Text)
+
+    # ComprehensiveEvaluation → User, MonthlyAchievement, CategoryProgress (N:1 관계)
+    user: Mapped["User"] = relationship(
+        "User", back_populates="comprehensive_evaluations", cascade="all, delete"
+    )
+    monthly_achievement: Mapped["MonthlyAchievement"] = relationship(
+        "MonthlyAchievement",
+        back_populates="comprehensive_evaluations",
+        cascade="all, delete",
+    )
+    category_progress: Mapped["CategoryProgress"] = relationship(
+        "CategoryProgress",
+        back_populates="comprehensive_evaluations",
+        cascade="all, delete",
```
```suggestion
    # ComprehensiveEvaluation → User, MonthlyAchievement, CategoryProgress (N:1 관계)
    user: Mapped["User"] = relationship(
        "User", back_populates="comprehensive_evaluations"
    )
    monthly_achievement: Mapped["MonthlyAchievement"] = relationship(
        "MonthlyAchievement",
        back_populates="comprehensive_evaluations"
    )
    category_progress: Mapped["CategoryProgress"] = relationship(
        "CategoryProgress",
        back_populates="comprehensive_evaluations"
```

codingorijazz commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,101 @@
+
+    # MonthlyAchievement → User (N:1 관계)
+    user: Mapped["User"] = relationship(
+        "User", back_populates="monthly_achievements", cascade="all, delete"
```
```suggestion
        "User", back_populates="monthly_achievements"
```

codingorijazz commented on app/models/evaluation.py:
```diff
@@ -0,0 +1,101 @@
+
+    # MonthlyAchievement → ComprehensiveEvaluation (1:N 관계)
+    comprehensive_evaluations: Mapped[List["ComprehensiveEvaluation"]] = relationship(
+        "ComprehensiveEvaluation", back_populates="monthly_achievements"
```
```suggestion
        "ComprehensiveEvaluation", back_populates="monthly_achievements", cascade="all, delete"
```

codingorijazz commented on app/models/user.py:
```diff
@@ -0,0 +1,86 @@
+
+    # User → UserCharacter, Goal, TodayNote, CategoryProgress, Chat, Storage, ComprehensiveEvaluation, MonthlyAchievement (1:N 관계)
+    user_characters: Mapped[List["UserCharacter"]] = relationship(
+        "UserCharacter", back_populates="users"
```
```suggestion
        "UserCharacter", back_populates="user", cascade="all, delete"
```

codingorijazz commented on app/models/user.py:
```diff
@@ -0,0 +1,86 @@
+
+    # UserCharacter → User, Character (N:1 관계)
+    user: Mapped["User"] = relationship(
+        "User", back_populates="user_characters", cascade="all, delete"
```
```suggestion
        "User", back_populates="user_characters"
```

codingorijazz commented on app/models/user.py:
```diff
@@ -0,0 +1,86 @@
+        "User", back_populates="user_characters", cascade="all, delete"
+    )
+    character: Mapped["Character"] = relationship(
+        "Character", back_populates="user_characters", cascade="all, delete"
```
```suggestion
        "Character", back_populates="user_characters"
```

codingorijazz commented on app/models/user.py:
```diff
@@ -0,0 +1,86 @@
+
+    # Character → UserCharacter (1:N 관계)
+    user_characters: Mapped[List["UserCharacter"]] = relationship(
+        "UserCharacter", back_populates="characters"
```
```suggestion
        "UserCharacter", back_populates="characters", cascade="all, delete"
```

codingorijazz reviewed (commented):
cascade="all, delete" 에 관한 내용입니다.

❗ cascade="all, delete"는 부모 → 자식 관계에서만 사용해야 합니다.
부모가 삭제될 때, 연결된 자식들도 함께 삭제되도록 설정하는 옵션입니다.
반대로 자식 → 부모 방향에서 이 옵션을 사용하는 것은 절대 금지입니다.
그렇게 하면 자식 데이터 하나 삭제할 때, 부모 데이터(예: User 전체)가 삭제될 수 있어 매우 위험합니다.

✅ 본 설정을 추가한 이유는 고아 레코드(Orphan record)를 방지하기 위함입니다.
예를 들어, MonthlyAchievement가 삭제되었는데 연결된 ComprehensiveEvaluation이 남아 있다면,
그 자식 데이터는 더 이상 유효하지 않은 쓰레기 데이터가 되어 데이터 무결성을 해치게 됩니다.

codingorijazz commented on app/models/category.py:
```diff
@@ -0,0 +1,102 @@
+
+    # Goal → User, CategoryProgresses (N : 1 관계)
+    user: Mapped["User"] = relationship(
+        "User", back_populates="goals", cascade="all, delete"
```
```suggestion
        "User", back_populates="goals"
```

codingorijazz commented on app/models/category.py:
```diff
@@ -0,0 +1,102 @@
+        "User", back_populates="goals", cascade="all, delete"
+    )
+    category_progress: Mapped["CategoryProgress"] = relationship(
+        "CategoryProgress", back_populates="goals", cascade="all, delete"
```
```suggestion
        "CategoryProgress", back_populates="goals"
```

codingorijazz commented on app/models/user.py:
```diff
@@ -0,0 +1,86 @@
+    user_characters: Mapped[List["UserCharacter"]] = relationship(
+        "UserCharacter", back_populates="users"
+    )
+    goals: Mapped[List["Goal"]] = relationship("Goal", back_populates="users")
```
```suggestion
    goals: Mapped[List["Goal"]] = relationship("Goal", back_populates="user", cascade="all, delete")
```

codingorijazz commented on app/models/category.py:
```diff
@@ -0,0 +1,102 @@
+        "RecommendedChallenge", back_populates="category_progresses"
+    )
+    goals: Mapped[List["Goal"]] = relationship(
+        "Goal", back_populates="category_progresses"
```
```suggestion
        "Goal", back_populates="category_progress", cascade="all, delete"
```

codingorijazz commented on app/models/category.py:
```diff
@@ -0,0 +1,102 @@
+
+    # TodayNote → User (N : 1 관계)
+    user: Mapped["User"] = relationship(
+        "User", back_populates="today_notes", cascade="all, delete"
```
```suggestion
        "User", back_populates="today_notes"
```

codingorijazz replied:
여기서 user로 변수명을 지정햇으니까, 
user.py에서 40번줄에 back_populates="user" 로 지정해야할 것 같습니다.

codingorijazz commented on app/models/user.py:
```diff
@@ -0,0 +1,86 @@
+    )
+    goals: Mapped[List["Goal"]] = relationship("Goal", back_populates="users")
+    today_notes: Mapped[List["TodayNote"]] = relationship(
+        "TodayNote", back_populates="users"
```
```suggestion
        "TodayNote", back_populates="user", cascade="all, delete"
```

codingorijazz replied:
category.py에 101번줄에서 user : 라고 변수명을 지정해서
여기에선 back_populates="user" 로 지정해야할 것 같습니다.

codingorijazz commented on app/models/user.py:
```diff
@@ -0,0 +1,86 @@
+        "CategoryProgress", back_populates="users"
+    )
+    chats: Mapped[List["Chat"]] = relationship("Chat", back_populates="users")
+    storages: Mapped[List["Storage"]] = relationship("Storage", back_populates="users")
```
```suggestion
    storages: Mapped[List["Storage"]] = relationship("Storage", back_populates="user", cascade="all, delete")
```

codingorijazz commented on app/models/chat.py:
```diff
@@ -0,0 +1,88 @@
+
+    # Function → Storage, Chat (1:N 관계)
+    storages: Mapped[List["Storage"]] = relationship(
+        "Storage", back_populates="functions"
```
```suggestion
        "Storage", back_populates="function"
```

codingorijazz commented on app/models/chat.py:
```diff
@@ -0,0 +1,88 @@
+    storages: Mapped[List["Storage"]] = relationship(
+        "Storage", back_populates="functions"
+    )
+    chats: Mapped[List["Chat"]] = relationship("Chat", back_populates="functions")
```
```suggestion
    chats: Mapped[List["Chat"]] = relationship("Chat", back_populates="function")
```

codingorijazz commented on app/models/category.py:
```diff
@@ -0,0 +1,102 @@
+        "ComprehensiveEvaluation", back_populates="category_progresses"
+    )
+    recommended_challenges: Mapped[List["RecommendedChallenge"]] = relationship(
+        "RecommendedChallenge", back_populates="category_progresses"
```
```suggestion
        "RecommendedChallenge", back_populates="category_progress", cascade="all, delete"
```

codingorijazz commented on app/models/user.py:
```diff
@@ -0,0 +1,86 @@
+    chats: Mapped[List["Chat"]] = relationship("Chat", back_populates="users")
+    storages: Mapped[List["Storage"]] = relationship("Storage", back_populates="users")
+    comprehensive_evaluations: Mapped[List["ComprehensiveEvaluation"]] = relationship(
+        "ComprehensiveEvaluation", back_populates="users"
```
```suggestion
        "ComprehensiveEvaluation", back_populates="user", cascade="all, delete"
```

codingorijazz commented on app/models/category.py:
```diff
@@ -0,0 +1,102 @@
+
+    # CategoryProgress → ComprehensiveEvaluation, RecommendedChallenge, Goal (1 : N 관계)
+    comprehensive_evaluations: Mapped[List["ComprehensiveEvaluation"]] = relationship(
+        "ComprehensiveEvaluation", back_populates="category_progresses"
```
```suggestion
        "ComprehensiveEvaluation", back_populates="category_progress", cascade="all, delete"
```

codingorijazz commented on app/models/user.py:
```diff
@@ -0,0 +1,86 @@
+        "ComprehensiveEvaluation", back_populates="users"
+    )
+    monthly_achievements: Mapped[List["MonthlyAchievement"]] = relationship(
+        "MonthlyAchievement", back_populates="users"
```
```suggestion
        "MonthlyAchievement", back_populates="user", cascade="all, delete"
```

codingorijazz commented on app/models/user.py:
```diff
@@ -0,0 +1,86 @@
+        "TodayNote", back_populates="users"
+    )
+    category_progresses: Mapped[List["CategoryProgress"]] = relationship(
+        "CategoryProgress", back_populates="users"
```
```suggestion
        "CategoryProgress", back_populates="user", cascade="all, delete"
```

codingorijazz commented on app/models/user.py:
```diff
@@ -0,0 +1,86 @@
+    category_progresses: Mapped[List["CategoryProgress"]] = relationship(
+        "CategoryProgress", back_populates="users"
+    )
+    chats: Mapped[List["Chat"]] = relationship("Chat", back_populates="users")
```
```suggestion
    chats: Mapped[List["Chat"]] = relationship("Chat", back_populates="user", cascade="all, delete")
```

Ryan-OH committed:
최종모델

Ryan-OH committed:
function 속성명 수정
